test/blackbox/BlackboxTestsLivelinessQos.cpp

```diff
@@ -0,0 +1,1424 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "BlackboxTests.hpp"
+
+#include "PubSubReader.hpp"
+#include "PubSubWriter.hpp"
+#include "PubSubParticipant.hpp"
+#include "ReqRepAsReliableHelloWorldRequester.hpp"
+#include "ReqRepAsReliableHelloWorldReplier.hpp"
+
+using namespace eprosima::fastrtps;
+using namespace eprosima::fastrtps::rtps;
+
+//! Tests that when kind is automatic liveliness is never lost, even if the writer never sends data
+TEST(LivelinessQos, Liveliness_Automatic_Reliable)
+{
+    PubSubReader<HelloWorldType> reader(TEST_TOPIC_NAME);
+    PubSubWriter<HelloWorldType> writer(TEST_TOPIC_NAME);
+
+    // Liveliness lease duration and announcement period
+    uint32_t liveliness_ms = 20;
+    Duration_t liveliness_s(liveliness_ms * 1e-3);
+    Duration_t announcement_period(liveliness_ms * 1e-3 * 0.9);
+
+    reader.reliability(RELIABLE_RELIABILITY_QOS)
+            .liveliness_kind(AUTOMATIC_LIVELINESS_QOS)
+            .liveliness_lease_duration(liveliness_s)
+            .init();
+    writer.reliability(RELIABLE_RELIABILITY_QOS)
+            .liveliness_kind(AUTOMATIC_LIVELINESS_QOS)
+            .liveliness_announcement_period(announcement_period)
+            .liveliness_lease_duration(liveliness_s)
+            .init();
+
+    ASSERT_TRUE(reader.isInitialized());
+    ASSERT_TRUE(writer.isInitialized());
+
+    // Wait for discovery.
+    writer.wait_discovery();
+    reader.wait_discovery();
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(liveliness_ms * 10));
+
+    // When using automatic kind, liveliness on both publisher and subscriber should never be lost
+    // It would only be lost if the publishing application crashed, which can't be reproduced in the tests
```

Comment:
You can perform this test using python. See test/communication/liveliness_assertion.py.

Reply:
Test added